server/src/main/java/io/crate/metadata/doc/DocTableInfoFactory.java

```diff
@@ -21,188 +21,660 @@
-            // We need all concrete indices, regardless of their state, for operations such as reopening.
+        IndexTemplateMetadata indexTemplateMetadata = metadata.templates().get(templateName);
+        Version versionCreated;
+        Version versionUpgraded;
+        Map<String, Object> mappingSource;
+        Settings tableParameters;
+        IndexMetadata.State state;
+        int numberOfShards;
+        String[] concreteIndices;
+        try {
             concreteIndices = IndexNameExpressionResolver.concreteIndexNames(
                 metadata,
-                IndicesOptions.lenientExpandOpen(),
+                indexTemplateMetadata == null
+                    ? IndicesOptions.strictExpandOpen()
+                    : IndicesOptions.lenientExpandOpen(),
                 relation.indexNameOrAlias()
             );
+        } catch (IndexNotFoundException e) {
+            throw new RelationUnknown(relation.fqn(), e);
+        }
+        String[] concreteOpenIndices;
+        List<PartitionName> partitions;
+        if (indexTemplateMetadata == null) {
+            IndexMetadata index = metadata.index(relation.indexNameOrAlias());
+            if (index == null) {
+                if (concreteIndices.length > 0) {
+                    index = metadata.index(concreteIndices[0]);
+                    LOGGER.info(
+                        "Found indices={} for relation={} without index template. Orphaned partition?",
+                        concreteIndices,
+                        relation
+                    );
+                }
+                if (index == null) {
+                    throw new RelationUnknown(relation);
+                }
+            }
+            Settings indexSettings = index.getSettings();
+            tableParameters = indexSettings;
+            versionCreated = IndexMetadata.SETTING_INDEX_VERSION_CREATED.get(indexSettings);
+            versionUpgraded = indexSettings.getAsVersion(IndexMetadata.SETTING_VERSION_UPGRADED, null);
+            state = index.getState();
+            MappingMetadata mapping = index.mapping();
+            mappingSource = mapping == null ? Map.of() : mapping.sourceAsMap();
+            numberOfShards = index.getNumberOfShards();
+            concreteOpenIndices = concreteIndices;
+            if (concreteIndices.length == 0) {
+                throw new RelationUnknown(relation);
+            }
+            partitions = List.of();
+        } else {
+            mappingSource = XContentHelper.toMap(
+                indexTemplateMetadata.mapping().compressedReference(),
+                XContentType.JSON
+            );
+            mappingSource = Maps.getOrDefault(mappingSource, "default", mappingSource);
+            Settings indexSettings = indexTemplateMetadata.settings();
+            tableParameters = indexSettings;
+            versionCreated = indexSettings.getAsVersion(IndexMetadata.SETTING_VERSION_CREATED, Version.CURRENT);
+            versionUpgraded = null;
```

Comment:
Don't we store the upgraded version also in templates?

Reply:
I thought templates can't be upgraded, only their partitions, but I'll double check

Reply:
I think this is the only place where it is upgraded:

https://github.com/crate/crate/blob/fefccf854e5b00cd8e062b2d0c6ff354c03a5c8b/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexUpgradeService.java#L186-L189

Operates on concrete `IndexMetadata`

Reply:
Recently stumbled over this and thought it's a bug.

Confirmed with https://github.com/crate/crate-qa/pull/332/commits, it fails with `AssertionError: None != '4.0.x'`

Probably related to scenario described in https://github.com/crate/crate/pull/17175#issue-2760665409.

I have a plan to upgrade not only template but indexMetadata as well - https://github.com/crate/crate/pull/17175#issuecomment-2563535411, basically, it looks like sometimes it's not enough to upgrade on startup.